src/ApiCompat/Microsoft.DotNet.ApiCompatibility/ApiComparer.cs

```diff
@@ -126,8 +144,26 @@ public class ApiComparer : IApiComparer
             IDifferenceVisitor visitor = _differenceVisitorFactory.Create();
             visitor.Visit(mapper);
 
-            // Sort the compat differences by the order of the passed in left assemblies.
-            return left.Join(visitor.CompatDifferences, i => i.MetadataInformation, d => d.Left, (i, d) => d).ToArray();
+            // Sort the compat differences by the order of the passed in lefts.
+            ILookup<MetadataInformation, CompatDifference> compatDifferenceLookup = visitor.CompatDifferences.ToLookup(d => d.Left, t => t);
+            HashSet<MetadataInformation> processedMetadata = new();
+            List<CompatDifference> compatDifferences = new();
+
+            foreach (ElementContainer<IAssemblySymbol> l in left)
+            {
+                compatDifferences.AddRange(compatDifferenceLookup[l.MetadataInformation]);
+                processedMetadata.Add(l.MetadataInformation);
+            }
+
+            foreach (IGrouping<MetadataInformation, CompatDifference> compatDifferenceGroup in compatDifferenceLookup)
+            {
+                if (processedMetadata.Contains(compatDifferenceGroup.Key))
+                    continue;
+
+                compatDifferences.AddRange(compatDifferenceGroup);
+            }
+
+            return compatDifferences;
```

Comment:
It seems you are doing the same logic for `right` and `left` to add them to `compatDifferences` and `processedMetadata`. Perhaps extract it into a function?